src/utils/page-functions.js

```diff
@@ -0,0 +1,202 @@
+      const body = await getPageFunctionBody(fnName, file, babelPresets)
+
+      return page.runAsyncScript({
+        args: [],
+        body: `${body} return (pageFn.${fnName} || pageFn.default).apply(null, arguments)`
+      }, ...args)
+    }
+  }
+
+  return pageFunctions
+}
+
+export async function getPageFunctionBody(fnName, filePath, presetOptions = {}) {
+  const cacheKey = createCacheKey(filePath, presetOptions)
+  const cacheFile = `${fnName}-${cacheKey}.js`
+  const cachePath = await getCacheEntry(cacheFile)
+  let cacheValid = await exists(cachePath)
+
+  if (cacheValid) {
+    const { mtime: fm } = await stats(filePath)
+    const { mtime: cm } = await stats(cachePath)
+
+    if (fm >= cm) {
+      cacheValid = false
+    }
+  }
+
+  if (!cacheValid) {
+    // transpile page function
+    await compilePageFunction({
+      name: 'pageFn',
+      filePath,
+      cachePath,
+      presetOptions
+    })
+  }
+
+  const fnBody = await readFile(cachePath, { encoding: 'utf8' })
+  return fnBody
+}
+
+export function compilePageFunction(config) {
+  return new Promise((resolve, reject) => {
+    const webpackConfig = createWebpackConfig(config)
+
+    webpack(webpackConfig, (err, stats) => {
+      /* istanbul ignore next */
+      if (err) {
+        reject(err)
+        return
+      }
+
+      if (stats.hasErrors()) {
+        const [err] = stats.compilation.errors
+        reject(err)
+        return
+      }
+
+      resolve(stats)
+    })
+  })
+}
+export function createWebpackConfig({
+  name,
+  filePath,
+  cachePath,
+  presetOptions
+}) {
+  return {
+    mode: 'production',
+    entry: filePath,
+    output: {
+      path: path.dirname(cachePath),
+      filename: path.basename(cachePath),
+      library: name,
+      libraryTarget: 'var'
+    },
+    module: {
+      rules: [
+        {
+          test: /\.js$/,
+          exclude: /node_modules/,
+          use: {
+            loader: 'babel-loader',
+            options: {
+              presets: [
+                ['@babel/preset-env', presetOptions]
+              ]
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+
+export function createCacheKey(fn, opts) {
+  if (typeof opts !== 'string') {
+    opts = JSON.stringify(opts)
+  }
+
+  return createHash('md5').update(`${fn}-x-${opts}`).digest('hex')
+}
+
+export async function getPackageName(modulesPath) {
+  if (!packageName) {
+    const { name } = await import(path.resolve(modulesPath, '..', 'package.json'))
+    packageName = name
+  }
+
+  return packageName
+}
+
+export async function getCacheEntry(filePath = '') {
+  const modulesPath = await findNodeModulesPath()
+  const name = await getPackageName(modulesPath)
+
+  return path.join(modulesPath, '.cache', name, filePath)
+}
+
+export async function findNodeModulesPath(refresh) {
+  if (nodeModulesPath && !refresh) {
+    return nodeModulesPath
+  }
+
+  const hablePath = requireResolve('hable/package.json')
```

Comment:
Why hable? :D

Reply:
I needed a file in node_modules that I knew would always exists, so I decided to use one of tib's own dependencies. hable seemed the best choice for that for various reasons, eg one reason because its probably the most 'unique' dependency but also just as a small hommage to nuxt's hook system ;)

Reply:
Hmm i didn't find out the usecase but if any other dep uses hable (another version), this will be broken. Why not checking package path itself?

Reply:
I was struggling to find a good reliable way of resolving the path to node_modules. Not sure what you mean with package path itself? Maybe process.cwd()?

Reply:
```js
path.resolve(__dirname, '..') // <-- this works for dist :)
```